### ISSUE TITLE
Cache Boogie translation and preprocessing at the Dafny module level

### ISSUE THREAD

**Issue by keyboardDrummer** [dafny-lang/dafny#2361]
Each Dafny module is translated to a separate Boogie program, which is then further processed by Boogie until Boogie discovers which verification tasks there are and to what extend they're already available in the cache. These Boogie preprocessing steps can be found here: https://github.com/boogie-org/boogie/blob/master/Source/ExecutionEngine/ExecutionEngine.cs#L684

### Change

- Cache the translation from a Dafny module to Boogie, and the subsequent Boogie processing steps that occur before Boogie's caching mechanism. The cache granularity will be a Dafny module, possibly including the dependencies of that module, in case those affect translation.

### Not in scope
- Caching at a higher granularity than modules such as classes or methods.
- Caching parsing, Dafny module resolution or verification
- Translating and Boogie preprocessing Dafny modules in parallel
- Changing the translation of a Dafny to a Boogie module

### Prerequisites
- https://github.com/dafny-lang/dafny/issues/2330

### Enables 
- https://github.com/dafny-lang/dafny/issues/2328
- https://github.com/dafny-lang/dafny/issues/2332

### Implementation hints
- Getting a module cache key should be easy after https://github.com/dafny-lang/dafny/issues/2330
- Cache entries can be removed if they were not used during a compilation
